apps: add initial support for hiding certain items from apis

phlogistonjohn opened the pull request:
Add the initial support for hiding certain items from api requests. Specifically, it adds the ability to hide volumes from the volume list, volume info, and cluster info requests. Nothing is actually hidden yet, this is just the framework.

raghavendra-talur commented:
I agree with most of the code as is. Just a question to make my understanding better : 

If I understand correctly, calling VolumeList() gives all volumes without filtering and using the lookupHandler version provides a filtered list. Why would you choose to implement of LookupHandler and not just add another VolumeListComplete() method on volume?

phlogistonjohn commented:
@raghavendra-talur : The next iteration of this code will need to filter out "pending" items. There are two ways to do this, looking at metadata inside the volume entry or looking at metadata on the pending operation entry. Doing the first means reading and parsing every volume entry, the latter means reading and parsing only the pending operations. I feel that in the average case there will be much fewer pending operations it should perform better, but to speed things up more I want to cache those pending operations once we parse them. So it seemed more appropriate to cache those entries in a common "lookup" object rather than in the Volume entry.

Secondarily, I feel like the volume entry objects are already kind of "overstuffed" with a lot of methods, whereas this new object is more single purpose.

obnoxxx commented on apps/glusterfs/lookup.go:
```diff
@@ -0,0 +1,45 @@
+
+// CompleteClusterInfo updates the given ClusterInfoResponse object so
+// that it only contains references to complete volumes, etc.
+func (lh *LookupHandler) CompleteClusterInfo(ci *api.ClusterInfoResponse) error {
```
This api is slightly asymmetric compared to the other ones.
The others just mimic the VolumeList and BrickList functions,
and return lists of objects, but this one does not mimic the clusterinfo
or so but instead potentially modifies the provided info structure.

Also the nomenclature is not quite consistent:

* CompleteVolumes returns the list of complete volumes
* CompleteBrick returns the list of complete bricks
* CompleteClusterInfo does not return the complete cluster info, but instead modifies the provided cluster info to only list the complete objects... 
 
Not entirely sure what to propose here, but it somehow feels a little odd.

obnoxxx commented on apps/glusterfs/app_cluster.go:
```diff
@@ -150,6 +150,7 @@ func (a *App) ClusterInfo(w http.ResponseWriter, r *http.Request) {
 		if err != nil {
 			return err
 		}
+		Lookup(tx).CompleteClusterInfo(info)
```
Even if it's currently just returning nil, we should still treat a possible future error condition here.

obnoxxx commented on apps/glusterfs/lookup_test.go:
```diff
@@ -0,0 +1,67 @@
+		tests.Assert(t, err == nil, "expected err == nil, got:", err)
+		info, err := ce.NewClusterInfoResponse(tx)
+		tests.Assert(t, err == nil, "expected err == nil, got:", err)
+		err = l.CompleteClusterInfo(info)
```
need to check error condition

obnoxxx commented:
Summary of agreed changes as a result of an offline discussion:
* For now, change this to a module that exposes 2 functions (for a start):
  * `ListVolumesComplete()`
  * `UpdateClusterInfoComplete()` (or so)
* I.e. remove the currently unneded BrickList variant (we can add it later if we want to)
* hide the Lookup api inside the module (we can later chose to expose it and /or change it to a callback model if we want to)
* add the return code checking of the Update call

Later we will add `ListVolumesPending()` and corresponding functions for BlockVolumes.

phlogistonjohn commented:
I've reworked the patches to provide a much simpler external interface: ListComplete(Block)Volumes & UpdateClusterInfoComplete. I tweaked the word order in the first two function names to be more natural when reading it. I also added the stubs for block volumes because I had done much of that work already.

obnoxxx commented:
@phlogistonjohn this needs rebase and I can't update :-)

phlogistonjohn commented:
Rebase'd

raghavendra-talur commented on apps/glusterfs/app_block_volume.go:
```diff
@@ -141,6 +141,9 @@ func (a *App) BlockVolumeInfo(w http.ResponseWriter, r *http.Request) {
 			return err
+		} else if !entry.Visible() {
+			http.Error(w, "Id not found", http.StatusNotFound)
+			return nil
```
i think we need to return ErrNotFound

phlogistonjohn replied:
I will double check it.

obnoxxx replied:
1.) it should treat the !visible condition just like not finding the entry. hence we need to return ErrNotFound
2.) In this case (as opposed to the volumeinfo function), the location is wrong. needs to in the block line 132 ff

raghavendra-talur commented on apps/glusterfs/app_volume.go:
```diff
@@ -207,6 +207,10 @@ func (a *App) VolumeInfo(w http.ResponseWriter, r *http.Request) {
+		} else if !entry.Visible() {
+			// treat an invisible entry like it doesn't exist
+			http.Error(w, "Id not found", http.StatusNotFound)
+			return nil
```
i think we need to return ErrNotFound

obnoxxx replied:
and merge with err == ErrNotFound condition?!

raghavendra-talur commented on apps/glusterfs/listing.go:
```diff
@@ -0,0 +1,35 @@
+
+// ListCompleteVolumes returns a list of volume ID strings for volumes
+// that are not pending.
+func ListCompleteVolumes(tx *bolt.Tx) ([]string, error) {
```
Just a suggestion, how about ListCompletedVolumes ?

phlogistonjohn replied:
That's fine too. I'll let @obnoxxx chime in with which version is preferred as a tiebreaker.

obnoxxx replied:
imho , let's stick with complete (1 char shorter...) ;-)

raghavendra-talur commented on apps/glusterfs/listing.go:
```diff
@@ -0,0 +1,35 @@
+
+// ListCompleteBlockVolumes returns a list of block volume ID strings for bricks
+// that are not pending.
+func ListCompleteBlockVolumes(tx *bolt.Tx) ([]string, error) {
```
same here, how about ListCompletedBlockVolumes ?

obnoxxx replied:
same, let's stick with ListCompleteBockVolumes, imho.

raghavendra-talur commented on apps/glusterfs/listing.go:
```diff
@@ -0,0 +1,35 @@
+	return BlockVolumeList(tx)
+}
+
+// CompleteClusterInfoComplete updates the given ClusterInfoResponse object so
```
I guess the comment needs to be updated.

phlogistonjohn replied:
Yeah, that just got messed up. That's neither the original name nor the current name. :-)

obnoxxx replied:
UpdateClusterInfoComplete it should be

obnoxxx commented:
@phlogistonjohn updated ... scary, that actually worked!

obnoxxx commented:
@phlogistonjohn btw, looking at and working on the code, the callback approach came to me again: we could also pass a filter function to `NewVolumeEntryFromId` and friends, or just create a filtering wrapper, like `NewCompleteVolumeEntryFromId` or so, that would include the Visible check, or so ... taking the need from the callers to do the check... 

well, something for the future.

obnoxxx commented:
For now, I think we've addressed all review comments

obnoxxx commented:
LGTM. requests have been addressed.

phlogistonjohn commented:
The updated changes LGTM
